test/auth_oidc/test_auth_oidc.py

```diff
@@ -252,104 +316,164 @@ def test_reauthenticate_succeeds(self):
         self.assertEqual(self.request_called, 3)
 
-    def test_reauthenticate_fails(self):
-        # Create request callback that returns valid credentials.
-        request_cb = self.create_request_cb()
-
-        # Create a client with the callback.
-        props: Dict = {"request_token_callback": request_cb}
-        client = MongoClient(self.uri_single, authmechanismproperties=props)
+        # Close the client.
+        client.close()
 
-        # Perform a find operation.
+    def test_4_4_reauthenticate_fails(self):
+        # Create a default OIDC client.
+        client = self.create_client()
+        # Perform a find operation that succeeds (to force a speculative auth).
         client.test.test.find_one()
-
-        # Assert that the request callback has been called once.
+        # Assert that the human callback has been called once.
         self.assertEqual(self.request_called, 1)
-
+        # Force a reauthentication using a failCommand.
         with self.fail_point(
             {
-                "mode": {"times": 2},
-                "data": {"failCommands": ["find"], "errorCode": 391},
+                "mode": {"times": 3},
+                "data": {"failCommands": ["find", "saslStart"], "errorCode": 391},
             }
         ):
             # Perform a find operation that fails.
             with self.assertRaises(OperationFailure):
                 client.test.test.find_one()
-
-        # Assert that the request callback has been called twice.
+        # Assert that the human callback has been called two times.
         self.assertEqual(self.request_called, 2)
+        # Close the client.
         client.close()
 
-    def test_reauthenticate_succeeds_bulk_write(self):
+    def test_request_callback_returns_null(self):
+        class RequestTokenNull(OIDCCallback):
+            def fetch(self, a):
+                return None
+
+        client = self.create_client(request_cb=RequestTokenNull())
+        with self.assertRaises(ValueError):
+            client.test.test.find_one()
+        client.close()
+
+    def test_request_callback_invalid_result(self):
+        class CallbackInvalidToken(OIDCCallback):
+            def fetch(self, a):
+                return {}
+
+        client = self.create_client(request_cb=CallbackInvalidToken())
+        with self.assertRaises(ValueError):
+            client.test.test.find_one()
+        client.close()
+
+    def test_reauthentication_succeeds_multiple_connections(self):
         request_cb = self.create_request_cb()
 
         # Create a client with the callback.
-        props: Dict = {"request_token_callback": request_cb}
-        client = MongoClient(self.uri_single, authmechanismproperties=props)
+        client1 = self.create_client(request_cb=request_cb)
+        client2 = self.create_client(request_cb=request_cb)
+
+        # Perform an insert operation.
+        client1.test.test.insert_many([{"a": 1}, {"a": 1}])
+        client2.test.test.find_one()
```

Comment:
Why do we not do an insert operation on client2 but still perform an operation? Is this just to trigger authentication for both?

Reply:
Yes